### PR TITLE
LSx oscillator is only initialized upon initial power up.

### PR DESCRIPTION
**Pull request by chuckbook**
This also includes VBAT!

If LSE is configured but fails to start, LSI is used until
next full power cycle.

Also handles STM32F7xx variant.
Reset flagging should be also removed (moved to modmachine).

### CODE REVIEW

File: stmhal/rtc.c [hunk 1]
```diff
@@ -178,49 +183,61 @@ void rtc_init(void) {
...
-        }
-        if(__HAL_RCC_GET_FLAG(RCC_FLAG_PINRST) != RESET) {
-            // external reset occurred
-            rtc_info |= 0x20000;
-        }
-        // Clear source Reset Flag
-        __HAL_RCC_CLEAR_RESET_FLAGS();
+    // fresh reset; configure RTC Calendar
+    RTC_CalendarConfig();
+    // RTC was previously set, so leave it alone
```
Comment (dpgeorge) [inline stmhal/rtc.c]: I think this comment is wrong/not needed.

File: stmhal/rtc.c [hunk 2]
```diff
@@ -178,49 +183,61 @@ void rtc_init(void) {
...
-        #else
-        PWR->CR |= PWR_CR_DBP;
-        #endif
+        HAL_PWR_EnableBkUpAccess();
         // Clear source Reset Flag
         __HAL_RCC_CLEAR_RESET_FLAGS();
         // provide some status information
         rtc_info |= 0x40000 | (RCC->BDCR & 7) | (RCC->CSR & 3) << 8;
         return;
+    } else if ((RCC->BDCR & RCC_BDCR_RTCSEL) == RCC_BDCR_RTCSEL_1) {
```
Comment (dpgeorge) [inline stmhal/rtc.c]: What does it mean if this condition succeeds?  Can there be a comment telling the state of the MCU if we get here?

**Comment by dpgeorge** [comment]
You spoke before of LSE "detection".  Did you mean what you implemented in this patch (that if LSE doesn't start, fallback to LSI), or can it be made more sophisticated whereby it only tries to init LSE if LSE is somehow detected as connected to the pins?

I've found that it has a very long timeout for LSE startup to fail, so takes ages to boot if LSE does not exist.

**Comment by chuckbook** [comment]
@dpgeorge 
First of all I have to admit that this my first trial of a push request. I'm still not convinced having done it the right way but my teacher ran away after throwing things at me :-) 
You are right, the comment on line 230 should be deleted.
Line 195 checks if LSI is already active.
The basic idea behind proposed RTC handling is to not initialize RTC if either LSE or LSI osc. is up and running.
Keep in mind that LSE could also be in bypass mode.
Initialization (including calendar) should be performed when no valid osc. was detected.
Checking for an installed crystal could be done but such a procedure would be very specific to a given MCU. I won't recommend it.
However, timeout for LSE detection is way too long. But that's what you defined initially.
(to be honest, without this boring delay I would have never touched this whole RTC thing -:)
According to my experience this timeout could be reduced to ~200ms, probably twice that under extreme temperature conditions.
With this timeout boot time would be extended by ~100ms in case of LSE failure.
Of course systems designed without LSE should disable it in mpconfig.h.
Although not strictly related to RTC initialization, I would like to bring in your idea of keeping pyb.freq() across resets. Some status has to survive a reset wherever that may be. So far one 32-bit word might be sufficient. Can you make a decision where to keep it?

**Comment by dpgeorge** [comment]
@chuckbook The timeout for LSE startup is defined by the ST HAL, and it's 5 seconds turning off then on the LSE.  I just assumed this was chosen by ST because it really needed to be that long.

> According to my experience this timeout could be reduced to ~200ms

From my experience programming nearly 6000 pyboards (with STM32F405RGT6), the average startup time for LSE is about 400ms, but can sometimes (around 2% of devices) extend to 600 or 700ms.

Do you know how to fix this?

> I would like to bring in your idea of keeping pyb.freq() across resets. Some status has to survive a reset wherever that may be. So far one 32-bit word might be sufficient. Can you make a decision where to keep it?

I think using the backup registers is the best option.  The backup SRAM is more flexible than the registers so should be left for the user to play with.  We can reserve, say, the last 32 backup registers (out of 70 or so) for our purposes and document that so people don't use them.

I think it's easiest and most efficient to just store the values of PLLM, PLLN, PLLP, PLLQ, and the AHB, APB1 and APB2 clock dividers, in the backup registers (maybe packing them together a bit so we don't need 7 registers), and then SystemClock_Config can just load those values directly.

**Comment by chuckbook** [comment]
@dpgeorge 
> Do you know how to fix this?
If we don't care about the ST HAL we can play some tricks, however, I'm not sure if this is the right approach. LSI fallback is considered to be an emergency path if the LSE is broken. A System with no LSE should use MICROPY_HW_RTC_USE_LSE. Don't you agree?
> I think it's easiest and most efficient to just store the values of PLLM, PLLN, ...
The freq solution I propose handles it exactly that way. All the info can be packed into one 32-bit word.
405 has 20 regs(80 bytes) others up to 32 (128 bytes). Can we reuse the old RTC_BKP_DR0?

**Comment by dpgeorge** [comment]
> If we don't care about the ST HAL we can play some tricks,

It would be nice to reduce the startup time of pyboard.  That's why I considered starting the LSE in the "background", but if you can make it quicker then that's a much better option.

> A System with no LSE should use MICROPY_HW_RTC_USE_LSE. Don't you agree?

Well, what about a board that has LSE as a hardware option, to be added by the user?  Then there would need to be 2 separate firmwares, one for no-LSE, one for user-installed-LSE.  Would be much cleaner to have 1 firmware that detected the LSE (this is a real use case for an upcoming variant of pyboard).

> The freq solution I propose handles it exactly that way. All the info can be packed into one 32-bit word.  405 has 20 regs(80 bytes) others up to 32 (128 bytes). Can we reuse the old RTC_BKP_DR0?

Ah, yes, it's 80 bytes, not 80 registers.  Let's use the last registers, so BKP19R on the F405 and BKP31R on F746, etc.  Then users are free to use the first few registers for their own purposes, independent of the board.

**Comment by peterhinch** [comment]
Has anyone looked with a scope to see why the oscillator takes so long to start up? High Q oscillators are inherently slow to start, but 700mS (~23000 cycles) seems excessive. Is it indicative of marginal gain? In which case I'd wonder if such a device would start at all at low temperature (as @chuckbook reports). Or is there a problem with the loading capacitor values or strays?

**Comment by blmorris** [comment]
@peterhinch - I don't have anything definitive, but like you I suspect it is because of the LSE driver's low feedback gain, one result of the low-power design. The signal on the LSE has a much lower amplitude than the HSE; this can be confirmed with a scope.
One of my first designs with this processor, the LSE would sometimes not start at all because the load capacitance was just a little too high; it took us few days to figure that one out...

**Comment by chuckbook** [comment]
> It would be nice to reduce the startup time of pyboard. That's why I considered starting the LSE in the "background", but if you can make it quicker then that's a much better option.

If we ignore the ST HAL we can start RTC in an optimistic way, meaning kick of RTC on initial power up. Time and date would be reset anyhow. A timer interrupt (few seconds) will verify LSE rdy and eventually fallback to LSI. All RTC related features have to be postponed.

> Well, what about a board that has LSE as a hardware option, to be added by the user? Then there would need to be 2 separate firmwares, one for no-LSE, one for user-installed-LSE. Would be much cleaner to have 1 firmware that detected the LSE (this is a real use case for an upcoming variant of pyboard).

Now I got the point, the "delayed" startup may handle this as well. If you plan to do LSE as an option you may consider one of the low power TXCO's.

I will use the top BKPxxR for freq handling.

**Comment by chuckbook** [comment]
@dpgeorge I stumbled together a first draft of the delayed rtc_init(). This is far from being a PR but it needs testing/discussion before more effort is put into it. How can we proceed?

**Comment by dpgeorge** [comment]
> If we ignore the ST HAL we can start RTC in an optimistic way, meaning kick of RTC on initial power up. Time and date would be reset anyhow. A timer interrupt (few seconds) will verify LSE rdy and eventually fallback to LSI. All RTC related features have to be postponed.
> We could even go without a timer interrupt and simply postpone the LSI decision until the first RTC utilization.

Yes, I like this.  I think go without a timer interrupt.  You'd need a check in all RTC functions that the RTC is working (and to start LSI if LSE failed).

> I stumbled together a first draft of the delayed rtc_init(). This is far from being a PR but it needs testing/discussion before more effort is put into it. How can we proceed?

Does your patch make this current PR obsolete, or does it build upon it?  If the latter then I'll merge this first.  Then you can just open another PR for your delayed rtc init and we can discuss the code there.

**Comment by chuckbook** [comment]
@dpgeorge If you don't mind PR 1571 should be merged first. Although I think I'm done with the delayed RTC stuff, I would like to complete power measurements with another minimal STM32F415 board. There is also a tiny chance to have a much faster LSE detector. It won't speed up LSE osc. init but skip LSE trial on an unpopulated crystal.

**Comment by chuckbook** [comment]
Finally finished the power measurements I mentioned before. For those interested in these things:
pyboard, nothing else connected, room temp.:
- idle USB REPL: ~23mA
- STANDBY: 31µA
- Power off, VBAR 3V RTC SS 256/s: 0.8µA
- Power off, VBAT 3V RTC SS 32768/s: 1.4µA
STM32F405 board with minimum parts populated, no LSE, HSE 24MHz, 3.3V power
unused pins flapping in the breeze:
- idle USB REPL: ~22mA
- STANDBY: 3.9µA (2.2µA at 2.7V power)
- Power off, VBAR 3V RTC SS 256/s: 0.8µA
- Power off, VBAT 3V RTC SS 32768/s: 1.4µA
BTW, same results after populating I2C pull-ups to Vdd. Could even operate one of our OLED modules which goes auto power down on STANDBY.
The firmware has delayed RTC initialization with LSI fallback. Typical boot time is ~100ms.
It also redefines datetime() sub-seconds field to increasing µs instead of decreasing 1/256s. 
@dpgeorge I managed to get in the LSE magic to skip LSE trial on unpopulated cystal/load caps.
However, this feature needs more testing with a broader sample base.

**Comment by dpgeorge** [comment]
This PR is now merged in 8f7ff854b0e341aa49d198a93fea7e83dd4cc71c, with minor style cleanups.

**Comment by dpgeorge** [comment]
@chuckbook thanks for the power measurements!  They agree with what I get, around 30uA for standby.  This is dominated by the ~25uA for the LDO leakage (PYBv1.1 is coming very soon and will use MCP1703 which has 2uA leakage and I measure ~6uA current in standby mode).

Can you please open another PR for the next set of improvements?  Thanks!

**Pull request closed without merging**